IPython/core/completer.py

```diff
@@ -2316,9 +2326,41 @@ def _jedi_matches(
             else:
                 return iter([])
 
+    @context_matcher()
+    def python_matcher(self, context: CompletionContext) -> SimpleMatcherResult:
+        """Match attributes or global python names"""
+        text = context.line_with_cursor
+        if "." in text:
+            try:
+                matches, fragment = self._attr_matches(text, include_prefix=False)
+                if text.endswith(".") and self.omit__names:
+                    if self.omit__names == 1:
+                        # true if txt is _not_ a __ name, false otherwise:
+                        no__name = lambda txt: re.match(r".*\.__.*?__", txt) is None
+                    else:
+                        # true if txt is _not_ a _ name, false otherwise:
+                        no__name = (
+                            lambda txt: re.match(r"\._.*?", txt[txt.rindex(".") :])
+                            is None
+                        )
+                    matches = filter(no__name, matches)
+                return _convert_matcher_v1_result_to_v2(
+                    matches, type="attribute", fragment=fragment
+                )
+            except NameError:
+                # catches <undefined attributes>.<tab>
+                matches = []
+                return _convert_matcher_v1_result_to_v2(matches, type="attribute")
+        else:
+            matches = self.global_matches(context.token)
+            return _convert_matcher_v1_result_to_v2(matches, type="variable")
```

Comment:
TODO: check the exact type name used by jedi to align if we can (is it attribute/variable or something different)?

Reply:
Jedi uses `statement`, LSP uses `variable`. I think `variable` is more appropriate here; a variable may however represent a more specialised "thing" such as `module` or `function`. While this is outside of this PR scope, I will add a comment to remember to update it.

Reply:
```suggestion
            # TODO: maybe distinguish between functions, modules and just "variables"
            return _convert_matcher_v1_result_to_v2(matches, type="variable")
```